[4.0] Feature: front-end users/user view

pe7er opened the pull request:
This PR add a new feature to **display a list of Users** from a **selected User Group**. You can use this new view for example to display a list of Users from the User Group "Author". 

### Summary of Changes
The front-end did not have a model/view to display a list of Users from a selected User Group. The back-end model of com_users has some authentication so that you cannot get data from the Users table if you are not logged in. 

This PR adds a Front-end View + Model for Users and User.
It adds a new menu item type In the back-end: 

### Testing Instructions
Create some Users that you assign to some User Group (e.g. Author and Editor).
Add some Articles that you assign to those new Users.

In Menus > Main Menu, create a New Menu Item of type "Users: List Users from a Group"

![users-backend1](https://user-images.githubusercontent.com/1217850/57194704-11e51680-6f4a-11e9-8f8a-b85670f548e5.png)

Under the "Users" tab, configure what the new menu item should display:
- User Groups: select a few
- Exclude Users: select a few
- Select if you want to see a Header or not
- Configure if you want to see the Number of Articles by that User
- Configure if you want to see the Last Visit date of that User
- Configure the Date Format if you want to show it differently
- Configure if you want to see the Online status of that User

![users-backend2](https://user-images.githubusercontent.com/1217850/57194703-11e51680-6f4a-11e9-89a0-daa46bdf97a1.png)

### Expected result

#### Front-end view: Users
![users-front-end1](https://user-images.githubusercontent.com/1217850/57194702-11e51680-6f4a-11e9-9d8b-589ef7442eaa.png)

#### Front-end view: Single User
![users-front-end2](https://user-images.githubusercontent.com/1217850/57194701-11e51680-6f4a-11e9-92a7-ace728bbb625.png)

pe7er committed:
feature: display list of Users on front-end - WIP

pe7er committed:
Front-end Users overview - WIP

pe7er committed:
User - Single view with articles - WIP

pe7er committed:
Users - list of articles - WIP

pe7er committed:
Users - list of users sortable

pe7er committed:
com_users - Front-end Users list overview + Single User view

pe7er committed:
com_users - Front-end Users list overview + Single User view

pe7er committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into feature-front-end-view-users

pe7er committed:
com_users - Front-end Users: Single User view - fixed article links

SniperSister commented on components/com_users/Controller/UserController.php:
```diff
@@ -360,4 +360,7 @@ public function resend()
 		// Check for request forgeries
 		// $this->checkToken('post');
 	}
+
```
I guess these blank lines can be removed

SniperSister commented on components/com_users/Model/UserModel.php:
```diff
@@ -0,0 +1,90 @@
+		$item = $db->loadObject();
+
+		// Get all Articles written by this Author
+		$item->articles = $this->getarticles($id);
```
Method name is camel case, so getArticles instead of getarticles. I also suggest to rename the method to i.e getUsersArticles with $userId in the method signature to make the naming more descriptive

pe7er replied:
Good point, thanks!

SniperSister commented on components/com_users/Model/UserModel.php:
```diff
@@ -0,0 +1,90 @@
+	public function getItem($id = null)
+	{
+		$db    = $this->getDbo();
+		$query = $db->getQuery(true)
```
Query should be refactored to use prepared statements

pe7er replied:
@SniperSister Ok, will do. 
Prepared statements were new for me but fortunately @wilsonge was here to explain :-)

SniperSister commented on components/com_users/Model/UsersModel.php:
```diff
@@ -0,0 +1,289 @@
+		$query = $db->getQuery(true);
+
+		// Select the required fields from the table.
+		$query->select(
```
Query should be refactored to use prepared statements

SniperSister commented on components/com_users/Model/UsersModel.php:
```diff
@@ -0,0 +1,289 @@
+	{
+		$app    = Factory::getApplication();
+		$params = ComponentHelper::getParams('com_users');
+		/*
```
Remove comment

SniperSister commented on components/com_users/tmpl/user/default.php:
```diff
@@ -0,0 +1,69 @@
+
+					<td headers="categorylist_header_title" class="list-title">
+						<a href="<?php echo Route::_('index.php?option=com_users&view=user&id=' . $article->id); ?>">
+							<?php echo $article->title; ?>
```
Article title needs escaping

SniperSister commented on components/com_users/tmpl/users/default.php:
```diff
@@ -0,0 +1,140 @@
+
+					<td headers="categorylist_header_title" class="list-title">
+						<a href="<?php echo Route::_('index.php?option=com_users&view=user&id=' . $item->id); ?>">
+							<?php echo $item->name; ?>
```
User name needs escaping

brianteeman replied:
This table cell should be a th

SniperSister commented:
General comment: we should make sure that by default, no listings are available. Site owners should be explicetly required to configure a menu item and select the corresponding user groups.

brianteeman commented:
on a site with 10,000 users is it practical to have the exclude users filter?

brianteeman commented:
In joomla 4 we don't use  class="chzn-color" instead it should be class="switcher"
and for the label it should not be the words "Show Feature" with options yes/no but simply "Feature" with options show/hide

brianteeman commented on components/com_users/tmpl/users/default.php:
```diff
@@ -0,0 +1,140 @@
+<div class="com-users-users user-list">
+	<form action="<?php echo Route::_('index.php?option=com_users&view=users'); ?>" method="post" name="adminForm" id="adminForm">
+		<table class="com-users-users__table users table table-striped table-bordered table-hover">
+			<?php if ($params->get('show_headings')) : ?>
```
all tables must have a caption
see https://github.com/joomla/joomla-cms/blob/bf1e0df5f97a6f6cc32843eb3e42431984d34df9/administrator/components/com_users/tmpl/users/default.php#L46-L48

pe7er replied:
@brianteeman Thanks!

pe7er committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into feature-front-end-view-users

SharkyKZ commented:
This sounds like a redo of #21441.

👎 for hardcoding `com_content` into this. Would be better if this was extendable with plugins.

brianteeman commented:
I agree with @SharkyKZ this should not be hardcoded to a single component

pe7er committed:
Refacored to Prepared Statements + Escaped some data + Added captionTable

pe7er commented:
> on a site with 10,000 users is it practical to have the exclude users filter?

Yes, in case you would have 10,000 Authors and you would exclude 3,000, 
then it is indeed rather impractical to select those 3,000 Users to exclude manually.

In that case I would advice you to create different User Groups (Authors_listed, Authors_unlisted).

This feature is useful if you just want to exclude a couple of Authors from being displayed.

brianteeman commented:
The 10,000 user list would still be creayed

pe7er commented:
> In joomla 4 we don't use class="chzn-color" instead it should be class="switcher"
> and for the label it should not be the words "Show Feature" with options yes/no but simply "Feature" with options show/hide

@brianteeman Ok, I have seen the class="switcher" being used for yes/no.

How would can I use it for 3 options?
For:  Use Global / Hide / Show ?

pe7er commented:
> The 10,000 user list would still be creayed

IMHO for the case I described, excluding a couple of users, the Exclude function would be workable

brianteeman commented:
> How would can I use it for 3 options?
You can't - it is only for 2. But I dont see any cases of you having 3 in this pr

brianteeman commented:
> > The 10,000 user list would still be creayed

> IMHO for the case I described, excluding a couple of users, the Exclude function would be workable

Sorry you missed my point. Its not about would the select list be usable it is that the select list is always going to be created. That will be a big performance hit

pe7er commented:
> This sounds like a redo of #21441.
> 
>  for hardcoding `com_content` into this. Would be better if this was extendable with plugins.

@SharkyKZ @brianteeman Feel free to improve this PR that it works with extendable plugins.

pe7er commented:
> Sorry you missed my point. Its not about would the select list be usable it is that the select list is always going to be created. That will be a big performance hit

@brianteeman Ok, you mean with the rendering of such a list itself... 
I thought you meant that selecting many users from the list would be difficult.

What would you recommend instead? 
A textarea where the userids numbers can be added comma delimited?

pe7er committed:
Changed labels

brianteeman commented:
I honestly don't know. I just knew it was an issue because I faced it on a site once before.

I am still not convinced that this is something for core

HLeithner commented on components/com_users/Model/UsersModel.php:
```diff
@@ -0,0 +1,282 @@
+		{
+			$groupIds = ArrayHelper::toInteger($groupIds);
+
+			// Todo: when Prepared Statements work with Arrays, add: ParameterType::INTEGER)
```
```suggestion
```

HLeithner commented on components/com_users/Model/UsersModel.php:
```diff
@@ -0,0 +1,282 @@
+		{
+			$excludedIds = ArrayHelper::toInteger($excludedIds);
+			$query->where(
+				$db->quoteName('users.id') . ' NOT IN (' . implode(',', $excludedIds) . ')'
```
```suggestion
				$db->quoteName('users.id') . ' NOT IN (' . implode(',', $db->bindArray($excludedIds)) . ')'
```

pe7er replied:
@HLeithner Thanks!

HLeithner commented on administrator/components/com_users/Field/UsersField.php:
```diff
@@ -0,0 +1,65 @@
+		$query = $db->getQuery(true)
+			->select(
+				$db->quoteName(
+					array(
```
please change to []

pe7er replied:
ok, thanks!

HLeithner commented on components/com_users/Model/UsersModel.php:
```diff
@@ -0,0 +1,282 @@
+	 *
+	 * @return  void
+	 * @throws \Exception
+	 * @since   1.6
```
wrong version

HLeithner committed:
Update components/com_users/Model/UsersModel.php

Co-Authored-By: pe7er <github@pe7er.com>

HLeithner committed:
Update components/com_users/Model/UsersModel.php

Co-Authored-By: pe7er <github@pe7er.com>

pe7er committed:
Merge branch '4.0-dev' into feature-front-end-view-users

brianteeman commented:
Looking forward to seeing the new and updated pr based on our discussion at #jd19nl

pe7er commented:
I've rewritten this PR to https://github.com/joomla/joomla-cms/pull/25030 in which I've decoupled it from the Articles. 
In the new PR I've also added Custom Fields support for Users.

Thank you all for testing + your feedback to this PR.
Closing this PR...

brianteeman commented:
Thanks for updating
